scripts/common/timer.ts

```diff
@@ -101,43 +96,329 @@ export interface RunSubsegment {
 	stats: RunStats;
 }
 
-export function getSplitSegmentTime(runSplits: RunSplits, segmentIndex: number, subsegmentIndex: number): number {
-	if (subsegmentIndex > 0) {
-		return (
-			runSplits.segments[segmentIndex].subsegments[subsegmentIndex].timeReached -
-			runSplits.segments[segmentIndex].subsegments[subsegmentIndex - 1].timeReached
-		);
+export interface RunStats {
+	maxOverallSpeed: float;
+	maxHorizontalSpeed: float;
+
+	overallDistanceTravelled: float;
+	horizontalDistanceTravelled: float;
+
+	jumps: uint16;
+	strafes: uint16;
+}
+
+export type RunStatType = keyof RunStats;
+
+export const RunStatsProperties: Record<RunStatType, { unit: string; isPositiveGood: boolean }> = {
+	maxOverallSpeed: {
+		unit: 'ups',
+		isPositiveGood: true
+	},
+	maxHorizontalSpeed: {
+		unit: 'ups',
+		isPositiveGood: true
+	},
+	overallDistanceTravelled: {
+		unit: 'units',
+		isPositiveGood: false
+	},
+	horizontalDistanceTravelled: {
+		unit: 'units',
+		isPositiveGood: false
+	},
+	jumps: {
+		unit: 'jumps',
+		isPositiveGood: false
+	},
+	strafes: {
+		unit: 'strafes',
+		isPositiveGood: false
+	}
+};
+
+//#endregion
+//#region Utility Functions
+
+/**
+ * For a given base and comparison splits, finds the split in base splits for a given majorNum and minorNum, then
+ * picks the appropriate split in the comparison splits, accounting for missing and unordered checkpoints.
+ */
+export function findSubsegmentComparison(
+	baseSplits: RunSplits,
+	compSplits: RunSplits | null,
+	majorNum: number,
+	minorNum: number
+): [RunSubsegment, RunSubsegment | null] {
+	const baseSeg = baseSplits.segments[majorNum - 1];
+	const baseSubsegIdx = baseSeg.subsegments.findIndex((ss) => ss.minorNum === minorNum);
+
+	// If this happens, something's gone wrong in the caller logic.
+	if (baseSubsegIdx === -1) {
+		throw new Error(`Missing subsegment for majorNum: ${majorNum}, minorNum: ${minorNum}`);
+	}
+
+	if (!compSplits) {
+		return [baseSeg.subsegments[baseSubsegIdx], null];
+	}
+
+	const compSeg = compSplits.segments[majorNum - 1];
```

Comment:
This shouldn't assume the comparison has segment `majorNum - 1`. I encountered this with with some comparison splits that were bugged on my end at the time, but this is something that could happen if zones differ (mostly a local zones scenario).

Reply:
Okay, returning `null` for comparison subseg if missing